edalize/flows/f4pga.py

```diff
@@ -0,0 +1,220 @@
+# Copyright edalize contributors
+# Licensed under the 2-Clause BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-2-Clause
+
+import os.path
+
+from edalize.flows.edaflow import Edaflow
+
+
+class F4pga(Edaflow):
+    """Edalize back-end implementation of F4PGA (Free and Open-Source Flow For FPGAs)"""
+
+    """Currently supports Xilinx 7-series chips"""
+
+    # F4PGA defined inputs to the underlying tools
+    FLOW_DEFINED_TOOL_OPTIONS = {}
+
+    # Inputs to the F4PGA flow
+    FLOW_OPTIONS = {
+        # Required (used by Yosys)
+        "device": {"type": "str", "desc": "Example: 'artix7'"},
+        # Required (used by Yosys)
+        "part": {"type": "str", "desc": "Example: 'xc7a35tcpg236-1'"},
+        # Required (used by VPR)
+        "chip": {"type": "str", "desc": "Example: 'xc7a50t_test'"},
+        # Optional, defaults to xilinx
+        "arch": {
+            "type": "str",
+            "desc": "Targeting architecture for Yosys. Currently supported is 'xilinx' and that is the default if none specified.",
+        },
+        # Optional, defaults to VPR
+        "pnr_tool": {
+            "type": "str",
+            "desc": "Place and route tool. Valid options are 'vpr'/'vtr' and 'nextpnr'. Defaults to VPR.",
+        },
+    }
+
+    # Standard F4PGA command line arguments to VPR
```

Comment:
This is a _lot_ of default options. Perhaps that's what's needed, but remember that it is usually harder to override a hardcoded default, because we need to keep backwards compatibility, so if any of these are commonly modified by the user, then I suggest just thinking through if they should be removed from the list.

Or perhaps vpr supports overriding an argument by specifying it twice on the command-line. In that case it's not a problem

Reply:
F4PGA sets all these options exactly, so I wasn't planning on allowing them to be configurable unless F4PGA decides to allow that. If you really think it's important I could write a system that would allow options to be passed in via tool_options or flow_options and then those would override the defaults, but I personally think that's a lot more effort than it would be worth right now.